gax/retry_policy_unittest.cc

```diff
@@ -0,0 +1,163 @@
+// Copyright 2019 Google Inc.  All rights reserved
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "googletest/include/gtest/gtest.h"
+#include "retry_policy.h"
+#include <chrono>
+#include <memory>
+
+namespace {
+using namespace ::google;
+
+class TestStatus {
+ public:
+  bool const isPermanent;
+};
+
+class TestRetryablePolicy {
+ public:
+  static bool IsPermanentFailure(TestStatus const& s) {
+    return s.isPermanent;
+  }
+};
+
+using RP = gax::RetryPolicy<TestStatus, TestRetryablePolicy>;
+
+using LECRP = gax::LimitedErrorCountRetryPolicy<TestStatus, TestRetryablePolicy>;
+
+TEST(LimitedErrorCountRetryPolicy, Basic) {
+  LECRP tested(3);
+  TestStatus s{false};
+  EXPECT_TRUE(tested.OnFailure(s));
+  EXPECT_TRUE(tested.OnFailure(s));
+  EXPECT_TRUE(tested.OnFailure(s));
+  EXPECT_FALSE(tested.OnFailure(s));
+  EXPECT_FALSE(tested.OnFailure(s));
+}
+
+TEST(LimitedErrorCountRetryPolicy, PermanentFailureCheck) {
+  LECRP tested(3);
+  TestStatus s{true};
+  EXPECT_FALSE(tested.OnFailure(s));
+}
+
+TEST(LimitedErrorCountRetryPolicy, CopyConstruct) {
+  LECRP tested(3);
+  TestStatus s{false};
+  EXPECT_TRUE(tested.OnFailure(s));
+  EXPECT_TRUE(tested.OnFailure(s));
+  EXPECT_TRUE(tested.OnFailure(s));
+  EXPECT_FALSE(tested.OnFailure(s));
+
+  LECRP copy(tested);
+  EXPECT_TRUE(copy.OnFailure(s));
+  EXPECT_TRUE(copy.OnFailure(s));
+  EXPECT_TRUE(copy.OnFailure(s));
+  EXPECT_FALSE(copy.OnFailure(s));
+}
+
+TEST(LimitedErrorCountRetryPolicy, MoveConstruct) {
```

Comment:
It isn't clear to me that this is the behaviour we want. It makes sense (I think...) for the copy constructor, but even then I'm not sure. But why should the retry policy forget the number of retries so far when it is moved?

Reply:
This is the current semantics of the reference implementation in google-cloud-cpp. It does make less sense for movement but does make sense for clone, and clone relies on the copy constructor.
I don't think I have a good idea of what common usage is going to look like. If all clients own and interact with policy objects by unique_ptr, and moving a unique_ptr can't trigger the move constructor of the wrapped object, then neither the move nor the copy constructor are ever used except internally.
The internal google-cloud-cpp Table client has a constructor that takes references to policies as parameters and assigns its policy attributes to clones of the parameters. It doesn't take ownership of the parameters. Is this similar to the workflow we want to use or encourage?

Reply:
Yes, we will have similar usage.

I guess I don't really object to this behaviour, I just don't understand why it is desirable. Let's keep this behaviour for now, since compatibility with google-cloud-cpp is our goal and there isn't a strong reason to change it.